// aws-amplify-react/dist/Amplify-UI/Amplify-UI-Components-React.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.propStyle = exports.beforeAfter = exports.PhotoPlaceholder = exports.Toast = exports.NavButton = exports.NavItem = exports.NavRight = exports.Nav = exports.NavBar = exports.InputLabel = exports.Hint = exports.Label = exports.Link = exports.SignInButtonContent = exports.SignInButtonIcon = exports.SignInButton = exports.PhotoPickerButton = exports.Button = exports.FormField = exports.SelectInput = exports.Input = exports.InputRow = exports.Radio = exports.RadioRow = exports.FormRow = exports.StrikeContent = exports.Strike = exports.ActionRow = exports.SectionBody = exports.SectionFooterSecondaryContent = exports.SectionFooterPrimaryContent = exports.SectionFooter = exports.SectionHeaderContent = exports.SectionHeader = exports.FormSection = exports.FormContainer = exports.Container = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

require('@aws-amplify/ui/dist/style.css');

var _ui = require('@aws-amplify/ui');

var AmplifyUI = _interopRequireWildcard(_ui);

var _AmplifyUITheme = require('./Amplify-UI-Theme');

var _AmplifyUITheme2 = _interopRequireDefault(_AmplifyUITheme);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Container = exports.Container = function Container(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.container);
    var p = _core.JS.objectLessAttributes(props, 'theme');

    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.container, style: style }),
        props.children
    ));
};

var FormContainer = exports.FormContainer = function FormContainer(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    return beforeAfter(_react2.default.createElement(
        'div',
        { className: AmplifyUI.formContainer, style: theme.formContainer },
        props.children
    ));
};

var FormSection = exports.FormSection = function FormSection(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = props.theme.formContainer || _AmplifyUITheme2.default.formContainer;
    return _react2.default.createElement(
        FormContainer,
        { theme: theme },
        beforeAfter(_react2.default.createElement(
            'div',
            { className: AmplifyUI.formSection, style: style },
            props.children
        ))
    );
};

var SectionHeader = exports.SectionHeader = function SectionHeader(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionHeader);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.sectionHeader, style: style }),
        _react2.default.createElement(
            SectionHeaderContent,
            { theme: theme },
            props.children,
            props.hint && _react2.default.createElement(
                'div',
                { className: AmplifyUI.sectionHeaderHint },
                props.hint
            )
        )
    ));
};

var SectionHeaderContent = exports.SectionHeaderContent = function SectionHeaderContent(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionHeaderContent);
    return beforeAfter(_react2.default.createElement(
        'span',
        { className: AmplifyUI.sectionHeaderContent, style: style },
        props.children
    ));
};

var SectionFooter = exports.SectionFooter = function SectionFooter(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionFooter);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.sectionFooter, style: style }),
        props.children
    ));
};

var SectionFooterPrimaryContent = exports.SectionFooterPrimaryContent = function SectionFooterPrimaryContent(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionFooterPrimaryContent);
    return beforeAfter(_react2.default.createElement(
        'span',
        { className: AmplifyUI.sectionFooterPrimaryContent, style: style },
        props.children
    ));
};

var SectionFooterSecondaryContent = exports.SectionFooterSecondaryContent = function SectionFooterSecondaryContent(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionFooterSecondaryContent);
    return beforeAfter(_react2.default.createElement(
        'span',
        { className: AmplifyUI.sectionFooterSecondaryContent, style: style },
        props.children
    ));
};

var SectionBody = exports.SectionBody = function SectionBody(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.sectionBody);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.sectionBody, style: style }),
        props.children
    ));
};

var ActionRow = exports.ActionRow = function ActionRow(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.actionRow);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: 'amplify-action-row', style: style }),
        props.children
    ));
};

var Strike = exports.Strike = function Strike(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.strike);
    return beforeAfter(_react2.default.createElement(
        'div',
        { className: AmplifyUI.strike, style: style },
        _react2.default.createElement(
            StrikeContent,
            { theme: theme },
            props.children
        )
    ));
};

var StrikeContent = exports.StrikeContent = function StrikeContent(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    return beforeAfter(_react2.default.createElement(
        'span',
        { className: AmplifyUI.strikeContent, style: theme.strikeContent },
        props.children
    ));
};

var FormRow = exports.FormRow = function FormRow(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.formRow);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.formRow, style: style }),
        props.children
    ));
};

var RadioRow = exports.RadioRow = function RadioRow(props) {
    var id = props.id || '_' + props.value;
    var theme = props.theme || _AmplifyUITheme2.default;
    return _react2.default.createElement(
        FormRow,
        { theme: theme },
        _react2.default.createElement(Radio, (0, _extends3.default)({}, props, {
            id: id
        })),
        _react2.default.createElement(
            Label,
            {
                htmlFor: id,
                theme: theme
            },
            props.placeholder
        )
    );
};

var Radio = exports.Radio = function Radio(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.radio);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement('input', (0, _extends3.default)({}, p, {
        type: 'radio',
        className: AmplifyUI.radio,
        style: style
    })));
};

var InputRow = exports.InputRow = function InputRow(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.input);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return _react2.default.createElement(
        FormRow,
        { theme: theme },
        beforeAfter(_react2.default.createElement('input', (0, _extends3.default)({}, p, { className: AmplifyUI.input, style: style })))
    );
};

var Input = exports.Input = function Input(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.input);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return _react2.default.createElement('input', (0, _extends3.default)({}, p, { className: AmplifyUI.input, style: style }));
};

var SelectInput = exports.SelectInput = function SelectInput(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.selectInput);
    return _react2.default.createElement(
        'div',
        { className: AmplifyUI.selectInput, style: style },
        props.children
    );
};

var FormField = exports.FormField = function FormField(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.formField);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.formField, style: style }),
        props.children
    ));
};

var Button = exports.Button = function Button(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.button);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'button',
        (0, _extends3.default)({}, p, { className: AmplifyUI.button, style: style }),
        props.children
    ));
};

var PhotoPickerButton = exports.PhotoPickerButton = function PhotoPickerButton(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.photoPickerButton);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return _react2.default.createElement(
        'button',
        (0, _extends3.default)({}, p, { className: [AmplifyUI.photoPickerButton, AmplifyUI.button].join(' '), style: style }),
        props.children
    );
};

var SignInButton = exports.SignInButton = function SignInButton(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.signInButton);
    var p = _core.JS.objectLessAttributes(props, 'theme');

    var classes = [AmplifyUI.signInButton];
    if (props.variant) classes.push(props.variant);

    return beforeAfter(_react2.default.createElement(
        'button',
        (0, _extends3.default)({}, p, { className: classes.join(' '), style: style }),
        props.children
    ));
};

var SignInButtonIcon = exports.SignInButtonIcon = function SignInButtonIcon(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.signInButtonIcon);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'span',
        (0, _extends3.default)({}, p, { className: AmplifyUI.signInButtonIcon, style: style }),
        props.children
    ));
};

var SignInButtonContent = exports.SignInButtonContent = function SignInButtonContent(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.signInButtonContent);
    return beforeAfter(_react2.default.createElement(
        'span',
        { className: AmplifyUI.signInButtonContent, style: style },
        props.children
    ));
};

var Link = exports.Link = function Link(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.a);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'a',
        (0, _extends3.default)({}, p, { className: AmplifyUI.a, style: style }),
        props.children
    ));
};

var Label = exports.Label = function Label(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.label);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'label',
        (0, _extends3.default)({}, p, { className: AmplifyUI.label, style: style }),
        props.children
    ));
};

var Hint = exports.Hint = function Hint(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.hint);
    return beforeAfter(_react2.default.createElement(
        'div',
        { className: AmplifyUI.hint, style: style },
        props.children
    ));
};

var InputLabel = exports.InputLabel = function InputLabel(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.inputLabel);
    return beforeAfter(_react2.default.createElement(
        'div',
        { className: AmplifyUI.inputLabel, style: style },
        props.children
    ));
};

var NavBar = exports.NavBar = function NavBar(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.navBar);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.navBar, style: style }),
        props.children
    ));
};

var Nav = exports.Nav = function Nav(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.nav);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.nav, style: style }),
        props.children
    ));
};

var NavRight = exports.NavRight = function NavRight(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.navRight);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.navRight, style: style }),
        props.children
    ));
};

var NavItem = exports.NavItem = function NavItem(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.navItem);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'div',
        (0, _extends3.default)({}, p, { className: AmplifyUI.navItem, style: style }),
        props.children
    ));
};

var NavButton = exports.NavButton = function NavButton(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.navButton);
    var p = _core.JS.objectLessAttributes(props, 'theme');
    return beforeAfter(_react2.default.createElement(
        'button',
        (0, _extends3.default)({}, p, { className: AmplifyUI.button, style: theme.toast }),
        beforeAfter(_react2.default.createElement(
            'span',
            null,
            props.children
        ))
    ));
};

var Toast = exports.Toast = function Toast(props) {
    var onClose = props.onClose;

    var theme = props.theme || _AmplifyUITheme2.default;

    return _react2.default.createElement(
        'div',
        { theme: theme, className: AmplifyUI.toast, style: theme.toast },
        _react2.default.createElement(
            'span',
            null,
            props.children
        ),
        _react2.default.createElement('a', { className: AmplifyUI.toastClose, onClick: onClose })
    );
};

Toast.defaultProps = {
    onClose: function onClose() {
        return void 0;
    }
};

var PhotoPlaceholder = exports.PhotoPlaceholder = function PhotoPlaceholder(props) {
    var theme = props.theme || _AmplifyUITheme2.default;
    var style = propStyle(props, theme.photoPlaceholder);
    return _react2.default.createElement(
        'div',
        { className: AmplifyUI.photoPlaceholder, style: style },
        _react2.default.createElement(
            'div',
            { className: AmplifyUI.photoPlaceholderIcon },
            _react2.default.createElement(
                'svg',
                { xmlns: 'http://www.w3.org/2000/svg', width: '64', height: '64', viewBox: '0 0 24 24' },
                _react2.default.createElement('circle', { cx: '12', cy: '12', r: '3.2' }),
                _react2.default.createElement('path', { d: 'M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z'
                }),
                _react2.default.createElement('path', { d: 'M0 0h24v24H0z', fill: 'none' })
            )
        )
    );
};

var beforeAfter = exports.beforeAfter = function beforeAfter(el) {
    var style = el.props.style || {};
    var before = style.before,
        after = style.after;

    if (!before && !after) {
        return el;
    }

    return _react2.default.createElement(
        'span',
        { style: { position: 'relative' } },
        before ? _react2.default.createElement(
            'span',
            { style: before },
            before.content
        ) : null,
        el,
        after ? _react2.default.createElement(
            'span',
            { style: after },
            after.content
        ) : null
    );
};

var propStyle = exports.propStyle = function propStyle(props, themeStyle) {
    var id = props.id,
        style = props.style;

    var styl = Object.assign({}, style, themeStyle);
    if (!id) {
        return styl;
    }

    var selector = '#' + id;
    Object.assign(styl, styl[selector]);
    return styl;
};
