// aws-amplify-react/dist/Amplify-UI/Amplify-UI-Theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

var Container = exports.Container = {};
var FormContainer = exports.FormContainer = {};
var FormSection = exports.FormSection = {};
var FormField = exports.FormField = {};
var SectionHeader = exports.SectionHeader = {};
var SectionBody = exports.SectionBody = {};
var SectionFooter = exports.SectionFooter = {};
var SectionFooterPrimaryContent = exports.SectionFooterPrimaryContent = {};
var SectionFooterSecondaryContent = exports.SectionFooterSecondaryContent = {};
var Input = exports.Input = {};
var Button = exports.Button = {};
var PhotoPickerButton = exports.PhotoPickerButton = {};
var PhotoPlaceholder = exports.PhotoPlaceholder = {};
var SignInButton = exports.SignInButton = {};
var SignInButtonIcon = exports.SignInButtonIcon = {};
var SignInButtonContent = exports.SignInButtonContent = {};
var Strike = exports.Strike = {};
var StrikeContent = exports.StrikeContent = {};
var ActionRow = exports.ActionRow = {};
var FormRow = exports.FormRow = {};
var A = exports.A = {};
var Hint = exports.Hint = {};
var Radio = exports.Radio = {};
var InputLabel = exports.InputLabel = {};

var AmplifyTheme = {
  container: Container,
  formContainer: FormContainer,
  formSection: FormSection,
  formField: FormField,

  sectionHeader: SectionHeader,
  sectionBody: SectionBody,
  sectionFooter: SectionFooter,
  sectionFooterPrimaryContent: SectionFooterPrimaryContent,
  sectionFooterSecondaryContent: SectionFooterSecondaryContent,

  input: Input,
  button: Button,
  photoPickerButton: PhotoPickerButton,
  photoPlaceholder: PhotoPlaceholder,
  signInButton: SignInButton,
  signInButtonIcon: SignInButtonIcon,
  signInButtonContent: SignInButtonContent,
  formRow: FormRow,
  strike: Strike,
  strikeContent: StrikeContent,
  actionRow: ActionRow,
  a: A,

  hint: Hint,
  radio: Radio,
  inputLabel: InputLabel
};

exports.default = AmplifyTheme;
