// aws-amplify-react/dist/Auth/AuthStateWrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

var _core2 = _interopRequireDefault(_core);

var _auth = require('@aws-amplify/auth');

var _auth2 = _interopRequireDefault(_auth);

var _AmplifyUITheme = require('../Amplify-UI/Amplify-UI-Theme');

var _AmplifyUITheme2 = _interopRequireDefault(_AmplifyUITheme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = new _core.ConsoleLogger('AuthStateWrapper');

var AuthStateWrapper = function (_Component) {
    (0, _inherits3.default)(AuthStateWrapper, _Component);

    function AuthStateWrapper(props) {
        (0, _classCallCheck3.default)(this, AuthStateWrapper);

        var _this = (0, _possibleConstructorReturn3.default)(this, (AuthStateWrapper.__proto__ || Object.getPrototypeOf(AuthStateWrapper)).call(this, props));

        _this.handleStateChange = _this.handleStateChange.bind(_this);
        _this.handleAuthEvent = _this.handleAuthEvent.bind(_this);
        _this.checkUser = _this.checkUser.bind(_this);

        _this.state = { auth: props.authState || 'signIn' };
        return _this;
    }

    (0, _createClass3.default)(AuthStateWrapper, [{
        key: 'componentWillMount',
        value: function componentWillMount() {
            var config = this.props.amplifyConfig;
            if (config) {
                _core2.default.configure(config);
            }
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.checkUser();
        }
    }, {
        key: 'handleStateChange',
        value: function handleStateChange(state, data) {
            logger.debug('authStateWrapper state change ' + state, data);
            if (state === this.state.auth) {
                return;
            }

            if (state === 'signedOut') {
                state = 'signIn';
            }
            this.setState({ auth: state, authData: data, error: null });
            if (this.props.onStateChange) {
                this.props.onStateChange(state, data);
            }
        }
    }, {
        key: 'handleAuthEvent',
        value: function handleAuthEvent(state, event) {
            if (event.type === 'error') {
                this.setState({ error: event.data });
            }
        }
    }, {
        key: 'checkUser',
        value: function checkUser() {
            var _this2 = this;

            if (!_auth2.default || typeof _auth2.default.currentUser !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }
            return _auth2.default.currentUser().then(function (user) {
                var state = user ? 'signedIn' : 'signIn';
                _this2.handleStateChange(state, user);
            }).catch(function (err) {
                return logger.error(err);
            });
        }
    }, {
        key: 'render',
        value: function render() {
            var _this3 = this;

            var _state = this.state,
                auth = _state.auth,
                authData = _state.authData;

            var theme = this.props.theme || _AmplifyUITheme2.default;
            var render_children = _react2.default.Children.map(this.props.children, function (child) {
                if (!child) {
                    return null;
                }
                return _react2.default.cloneElement(child, {
                    authState: auth,
                    authData: authData,
                    theme: theme,
                    onStateChange: _this3.handleStateChange,
                    onAuthEvent: _this3.handleAuthEvent
                });
            });

            return _react2.default.createElement(
                'div',
                { className: 'amplify-state-wrapper', style: theme.stateWrapper },
                render_children,
                this.state.error && _react2.default.createElement(
                    'div',
                    {
                        className: 'amplify-error-section',
                        style: theme.errorSection
                    },
                    this.state.error
                )
            );
        }
    }]);
    return AuthStateWrapper;
}(_react.Component);

exports.default = AuthStateWrapper;
