// aws-amplify-react/dist/Auth/Authenticator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

var _auth = require('@aws-amplify/auth');

var _auth2 = _interopRequireDefault(_auth);

var _Greetings = require('./Greetings');

var _Greetings2 = _interopRequireDefault(_Greetings);

var _SignIn = require('./SignIn');

var _SignIn2 = _interopRequireDefault(_SignIn);

var _ConfirmSignIn = require('./ConfirmSignIn');

var _ConfirmSignIn2 = _interopRequireDefault(_ConfirmSignIn);

var _RequireNewPassword = require('./RequireNewPassword');

var _RequireNewPassword2 = _interopRequireDefault(_RequireNewPassword);

var _SignUp = require('./SignUp');

var _SignUp2 = _interopRequireDefault(_SignUp);

var _ConfirmSignUp = require('./ConfirmSignUp');

var _ConfirmSignUp2 = _interopRequireDefault(_ConfirmSignUp);

var _VerifyContact = require('./VerifyContact');

var _VerifyContact2 = _interopRequireDefault(_VerifyContact);

var _ForgotPassword = require('./ForgotPassword');

var _ForgotPassword2 = _interopRequireDefault(_ForgotPassword);

var _TOTPSetup = require('./TOTPSetup');

var _TOTPSetup2 = _interopRequireDefault(_TOTPSetup);

var _AmplifyUITheme = require('../Amplify-UI/Amplify-UI-Theme');

var _AmplifyUITheme2 = _interopRequireDefault(_AmplifyUITheme);

var _AmplifyMessageMap = require('../AmplifyMessageMap');

var _AmplifyMessageMap2 = _interopRequireDefault(_AmplifyMessageMap);

var _AmplifyUIComponentsReact = require('../Amplify-UI/Amplify-UI-Components-React');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = new _core.ConsoleLogger('Authenticator'); /*
                                                        * Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
                                                        *
                                                        * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
                                                        * the License. A copy of the License is located at
                                                        *
                                                        *     http://aws.amazon.com/apache2.0/
                                                        *
                                                        * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
                                                        * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
                                                        * and limitations under the License.
                                                        */

var Authenticator = function (_Component) {
    (0, _inherits3.default)(Authenticator, _Component);

    function Authenticator(props) {
        (0, _classCallCheck3.default)(this, Authenticator);

        var _this = (0, _possibleConstructorReturn3.default)(this, (Authenticator.__proto__ || Object.getPrototypeOf(Authenticator)).call(this, props));

        _this.handleStateChange = _this.handleStateChange.bind(_this);
        _this.handleAuthEvent = _this.handleAuthEvent.bind(_this);
        _this.onHubCapsule = _this.onHubCapsule.bind(_this);

        _this.state = { auth: props.authState || 'loading' };
        _core.Hub.listen('auth', _this);
        return _this;
    }

    (0, _createClass3.default)(Authenticator, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            var config = this.props.amplifyConfig;
            if (config) {
                _core.Amplify.configure(config);
            }
            this._isMounted = true;
            this.checkUser();
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            this._isMounted = false;
        }
    }, {
        key: 'checkUser',
        value: function checkUser() {
            var _this2 = this;

            var auth = this.state.auth;

            if (!_auth2.default || typeof _auth2.default.currentAuthenticatedUser !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }
            return _auth2.default.currentAuthenticatedUser().then(function (user) {
                if (!_this2._isMounted) {
                    return;
                }
                if (auth !== 'signedIn') {
                    _this2.setState({
                        authState: 'signedIn',
                        authData: user
                    });
                    _this2.handleStateChange('signedIn', user);
                }
            }).catch(function (err) {
                if (!_this2._isMounted) {
                    return;
                }
                _auth2.default.signOut().then(function () {
                    return _this2.handleStateChange('signedOut');
                });
            });
        }
    }, {
        key: 'onHubCapsule',
        value: function onHubCapsule(capsule) {
            var channel = capsule.channel,
                payload = capsule.payload,
                source = capsule.source;

            if (channel === 'auth' && (payload.event === 'configured' || payload.event === 'cognitoHostedUI')) {
                this.checkUser();
            }
        }
    }, {
        key: 'handleStateChange',
        value: function handleStateChange(state, data) {
            logger.debug('authenticator state change ' + state, data);
            if (state === this.state.auth) {
                return;
            }

            if (state === 'signedOut') {
                state = 'signIn';
            }
            this.setState({ auth: state, authData: data, error: null });
            if (this.props.onStateChange) {
                this.props.onStateChange(state, data);
            }
        }
    }, {
        key: 'handleAuthEvent',
        value: function handleAuthEvent(state, event) {
            if (event.type === 'error') {
                var map = this.props.errorMessage || _AmplifyMessageMap2.default;
                var message = typeof map === 'string' ? map : map(event.data);
                this.setState({ error: message, showToast: true });
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _this3 = this;

            var _state = this.state,
                auth = _state.auth,
                authData = _state.authData;

            var theme = this.props.theme || _AmplifyUITheme2.default;
            var messageMap = this.props.errorMessage || _AmplifyMessageMap2.default;

            var _props = this.props,
                hideDefault = _props.hideDefault,
                _props$hide = _props.hide,
                hide = _props$hide === undefined ? [] : _props$hide,
                federated = _props.federated;

            if (hideDefault) {
                hide = hide.concat([_Greetings2.default, _SignIn2.default, _ConfirmSignIn2.default, _RequireNewPassword2.default, _SignUp2.default, _ConfirmSignUp2.default, _VerifyContact2.default, _ForgotPassword2.default, _TOTPSetup2.default]);
            }
            var props_children = this.props.children || [];

            var default_children = [_react2.default.createElement(_Greetings2.default, null), _react2.default.createElement(_SignIn2.default, { federated: federated }), _react2.default.createElement(_ConfirmSignIn2.default, null), _react2.default.createElement(_RequireNewPassword2.default, null), _react2.default.createElement(_SignUp2.default, null), _react2.default.createElement(_ConfirmSignUp2.default, null), _react2.default.createElement(_VerifyContact2.default, null), _react2.default.createElement(_ForgotPassword2.default, null), _react2.default.createElement(_TOTPSetup2.default, null)];

            var props_children_names = _react2.default.Children.map(props_children, function (child) {
                return child.type.name;
            });
            hide = hide.filter(function (component) {
                return !props_children_names.includes(component.name);
            });
            var render_props_children = _react2.default.Children.map(props_children, function (child, index) {
                return _react2.default.cloneElement(child, {
                    key: 'aws-amplify-authenticator-props-children-' + index,
                    theme: theme,
                    messageMap: messageMap,
                    authState: auth,
                    authData: authData,
                    onStateChange: _this3.handleStateChange,
                    onAuthEvent: _this3.handleAuthEvent,
                    hide: hide
                });
            });

            var render_default_children = hideDefault ? [] : _react2.default.Children.map(default_children, function (child, index) {
                return _react2.default.cloneElement(child, {
                    key: 'aws-amplify-authenticator-default-children-' + index,
                    theme: theme,
                    messageMap: messageMap,
                    authState: auth,
                    authData: authData,
                    onStateChange: _this3.handleStateChange,
                    onAuthEvent: _this3.handleAuthEvent,
                    hide: hide
                });
            });

            var render_children = render_default_children.concat(render_props_children);
            var error = this.state.error;

            return _react2.default.createElement(
                _AmplifyUIComponentsReact.Container,
                { theme: theme },
                this.state.showToast && _react2.default.createElement(
                    _AmplifyUIComponentsReact.Toast,
                    { onClose: function onClose() {
                            return _this3.setState({ showToast: false });
                        } },
                    _core.I18n.get(error)
                ),
                render_children
            );
        }
    }]);
    return Authenticator;
}(_react.Component);

exports.default = Authenticator;
