// aws-amplify-react/dist/Auth/ConfirmSignUp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

var _auth = require('@aws-amplify/auth');

var _auth2 = _interopRequireDefault(_auth);

var _AuthPiece2 = require('./AuthPiece');

var _AuthPiece3 = _interopRequireDefault(_AuthPiece2);

var _AmplifyUIComponentsReact = require('../Amplify-UI/Amplify-UI-Components-React');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = new _core.ConsoleLogger('ConfirmSignUp'); /*
                                                        * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
                                                        *
                                                        * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
                                                        * the License. A copy of the License is located at
                                                        *
                                                        *     http://aws.amazon.com/apache2.0/
                                                        *
                                                        * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
                                                        * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
                                                        * and limitations under the License.
                                                        */

var ConfirmSignUp = function (_AuthPiece) {
    (0, _inherits3.default)(ConfirmSignUp, _AuthPiece);

    function ConfirmSignUp(props) {
        (0, _classCallCheck3.default)(this, ConfirmSignUp);

        var _this = (0, _possibleConstructorReturn3.default)(this, (ConfirmSignUp.__proto__ || Object.getPrototypeOf(ConfirmSignUp)).call(this, props));

        _this._validAuthStates = ['confirmSignUp'];
        _this.confirm = _this.confirm.bind(_this);
        _this.resend = _this.resend.bind(_this);
        return _this;
    }

    (0, _createClass3.default)(ConfirmSignUp, [{
        key: 'confirm',
        value: function confirm() {
            var _this2 = this;

            var username = this.usernameFromAuthData() || this.inputs.username;
            var code = this.inputs.code;

            if (!_auth2.default || typeof _auth2.default.confirmSignUp !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }

            _auth2.default.confirmSignUp(username, code).then(function () {
                return _this2.changeState('signedUp');
            }).catch(function (err) {
                return _this2.error(err);
            });
        }
    }, {
        key: 'resend',
        value: function resend() {
            var _this3 = this;

            var username = this.usernameFromAuthData() || this.inputs.username;
            if (!_auth2.default || typeof _auth2.default.resendSignUp !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }
            _auth2.default.resendSignUp(username).then(function () {
                return logger.debug('code resent');
            }).catch(function (err) {
                return _this3.error(err);
            });
        }
    }, {
        key: 'showComponent',
        value: function showComponent(theme) {
            var _this4 = this;

            var hide = this.props.hide;

            var username = this.usernameFromAuthData();

            if (hide && hide.includes(ConfirmSignUp)) {
                return null;
            }

            return _react2.default.createElement(
                _AmplifyUIComponentsReact.FormSection,
                { theme: theme },
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionHeader,
                    { theme: theme },
                    _core.I18n.get('Confirm Sign Up')
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionBody,
                    { theme: theme },
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.FormField,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.InputLabel,
                            null,
                            _core.I18n.get('Username'),
                            ' *'
                        ),
                        _react2.default.createElement(_AmplifyUIComponentsReact.Input, {
                            placeholder: _core.I18n.get('Username'),
                            theme: theme,
                            key: 'username',
                            name: 'username',
                            onChange: this.handleInputChange,
                            disabled: username,
                            value: username ? username : ""
                        })
                    ),
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.FormField,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.InputLabel,
                            null,
                            _core.I18n.get('Confirmation Code'),
                            ' *'
                        ),
                        _react2.default.createElement(_AmplifyUIComponentsReact.Input, {
                            autoFocus: true,
                            placeholder: _core.I18n.get('Enter your code'),
                            theme: theme,
                            key: 'code',
                            name: 'code',
                            autoComplete: 'off',
                            onChange: this.handleInputChange
                        }),
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Hint,
                            { theme: theme },
                            _core.I18n.get('Lost your code? '),
                            _react2.default.createElement(
                                _AmplifyUIComponentsReact.Link,
                                { theme: theme, onClick: this.resend },
                                _core.I18n.get('Resend Code')
                            )
                        )
                    )
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionFooter,
                    { theme: theme },
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterPrimaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Button,
                            { theme: theme, onClick: this.confirm },
                            _core.I18n.get('Confirm')
                        )
                    ),
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterSecondaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Link,
                            { theme: theme, onClick: function onClick() {
                                    return _this4.changeState('signIn');
                                } },
                            _core.I18n.get('Back to Sign In')
                        )
                    )
                )
            );
        }
    }]);
    return ConfirmSignUp;
}(_AuthPiece3.default);

exports.default = ConfirmSignUp;
