// aws-amplify-react/dist/Auth/ConfirmSignIn.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

var _auth = require('@aws-amplify/auth');

var _auth2 = _interopRequireDefault(_auth);

var _AuthPiece2 = require('./AuthPiece');

var _AuthPiece3 = _interopRequireDefault(_AuthPiece2);

var _AmplifyUIComponentsReact = require('../Amplify-UI/Amplify-UI-Components-React');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = new _core.ConsoleLogger('ConfirmSignIn'); /*
                                                        * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
                                                        *
                                                        * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
                                                        * the License. A copy of the License is located at
                                                        *
                                                        *     http://aws.amazon.com/apache2.0/
                                                        *
                                                        * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
                                                        * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
                                                        * and limitations under the License.
                                                        */

var ConfirmSignIn = function (_AuthPiece) {
    (0, _inherits3.default)(ConfirmSignIn, _AuthPiece);

    function ConfirmSignIn(props) {
        (0, _classCallCheck3.default)(this, ConfirmSignIn);

        var _this = (0, _possibleConstructorReturn3.default)(this, (ConfirmSignIn.__proto__ || Object.getPrototypeOf(ConfirmSignIn)).call(this, props));

        _this._validAuthStates = ['confirmSignIn'];
        _this.confirm = _this.confirm.bind(_this);
        _this.checkContact = _this.checkContact.bind(_this);
        _this.state = {
            mfaType: 'SMS'
        };
        return _this;
    }

    (0, _createClass3.default)(ConfirmSignIn, [{
        key: 'checkContact',
        value: function checkContact(user) {
            var _this2 = this;

            if (!_auth2.default || typeof _auth2.default.verifiedContact !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }

            _auth2.default.verifiedContact(user).then(function (data) {
                if (!_core.JS.isEmpty(data.verified)) {
                    _this2.changeState('signedIn', user);
                } else {
                    user = Object.assign(user, data);
                    _this2.changeState('verifyContact', user);
                }
            });
        }
    }, {
        key: 'confirm',
        value: function confirm() {
            var _this3 = this;

            var user = this.props.authData;
            var code = this.inputs.code;

            var mfaType = user.challengeName === 'SOFTWARE_TOKEN_MFA' ? 'SOFTWARE_TOKEN_MFA' : null;
            if (!_auth2.default || typeof _auth2.default.confirmSignIn !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }

            _auth2.default.confirmSignIn(user, code, mfaType).then(function () {
                _this3.checkContact(user);
            }).catch(function (err) {
                return _this3.error(err);
            });
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
            //logger.debug('component did update with props', this.props);
            var user = this.props.authData;
            var mfaType = user && user.challengeName === 'SOFTWARE_TOKEN_MFA' ? 'TOTP' : 'SMS';
            if (this.state.mfaType !== mfaType) this.setState({ mfaType: mfaType });
        }
    }, {
        key: 'showComponent',
        value: function showComponent(theme) {
            var _this4 = this;

            var _props = this.props,
                hide = _props.hide,
                authData = _props.authData;

            if (hide && hide.includes(ConfirmSignIn)) {
                return null;
            }

            return _react2.default.createElement(
                _AmplifyUIComponentsReact.FormSection,
                { theme: theme },
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionHeader,
                    { theme: theme },
                    _core.I18n.get('Confirm ' + this.state.mfaType + ' Code')
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionBody,
                    { theme: theme },
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.FormField,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.InputLabel,
                            null,
                            _core.I18n.get('Code'),
                            ' *'
                        ),
                        _react2.default.createElement(_AmplifyUIComponentsReact.Input, {
                            autoFocus: true,
                            placeholder: _core.I18n.get('Code'),
                            theme: theme,
                            key: 'code',
                            name: 'code',
                            autoComplete: 'off',
                            onChange: this.handleInputChange
                        })
                    )
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionFooter,
                    { theme: theme },
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterPrimaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Button,
                            { theme: theme, onClick: this.confirm },
                            _core.I18n.get('Confirm')
                        )
                    ),
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterSecondaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Link,
                            { theme: theme, onClick: function onClick() {
                                    return _this4.changeState('signIn');
                                } },
                            _core.I18n.get('Back to Sign In')
                        )
                    )
                )
            );
        }
    }]);
    return ConfirmSignIn;
}(_AuthPiece3.default);

exports.default = ConfirmSignIn;
