// aws-amplify-react/dist/Auth/RequireNewPassword.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _core = require('@aws-amplify/core');

var _auth = require('@aws-amplify/auth');

var _auth2 = _interopRequireDefault(_auth);

var _AuthPiece2 = require('./AuthPiece');

var _AuthPiece3 = _interopRequireDefault(_AuthPiece2);

var _AmplifyTheme = require('../AmplifyTheme');

var _AmplifyTheme2 = _interopRequireDefault(_AmplifyTheme);

var _AmplifyUIComponentsReact = require('../Amplify-UI/Amplify-UI-Components-React');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

var logger = new _core.ConsoleLogger('RequireNewPassword');

var RequireNewPassword = function (_AuthPiece) {
    (0, _inherits3.default)(RequireNewPassword, _AuthPiece);

    function RequireNewPassword(props) {
        (0, _classCallCheck3.default)(this, RequireNewPassword);

        var _this = (0, _possibleConstructorReturn3.default)(this, (RequireNewPassword.__proto__ || Object.getPrototypeOf(RequireNewPassword)).call(this, props));

        _this._validAuthStates = ['requireNewPassword'];
        _this.change = _this.change.bind(_this);
        _this.checkContact = _this.checkContact.bind(_this);
        return _this;
    }

    (0, _createClass3.default)(RequireNewPassword, [{
        key: 'checkContact',
        value: function checkContact(user) {
            var _this2 = this;

            if (!_auth2.default || typeof _auth2.default.verifiedContact !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }
            _auth2.default.verifiedContact(user).then(function (data) {
                if (!_core.JS.isEmpty(data.verified)) {
                    _this2.changeState('signedIn', user);
                } else {
                    user = Object.assign(user, data);
                    _this2.changeState('verifyContact', user);
                }
            });
        }
    }, {
        key: 'change',
        value: function change() {
            var _this3 = this;

            var user = this.props.authData;
            var password = this.inputs.password;
            var requiredAttributes = user.challengeParam.requiredAttributes;

            var attrs = objectWithProperties(this.inputs, requiredAttributes);

            if (!_auth2.default || typeof _auth2.default.completeNewPassword !== 'function') {
                throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
            }
            _auth2.default.completeNewPassword(user, password, attrs).then(function (user) {
                logger.debug('complete new password', user);
                if (user.challengeName === 'SMS_MFA') {
                    _this3.changeState('confirmSignIn', user);
                } else if (user.challengeName === 'MFA_SETUP') {
                    logger.debug('TOTP setup', user.challengeParam);
                    _this3.changeState('TOTPSetup', user);
                } else {
                    _this3.checkContact(user);
                }
            }).catch(function (err) {
                return _this3.error(err);
            });
        }
    }, {
        key: 'showComponent',
        value: function showComponent(theme) {
            var _this4 = this;

            var hide = this.props.hide;

            if (hide && hide.includes(RequireNewPassword)) {
                return null;
            }

            var user = this.props.authData;
            var requiredAttributes = user.challengeParam.requiredAttributes;


            return _react2.default.createElement(
                _AmplifyUIComponentsReact.FormSection,
                { theme: theme },
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionHeader,
                    { theme: theme },
                    _core.I18n.get('Change Password')
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionBody,
                    null,
                    _react2.default.createElement(_AmplifyUIComponentsReact.Input, {
                        autoFocus: true,
                        placeholder: _core.I18n.get('New Password'),
                        theme: theme,
                        key: 'password',
                        name: 'password',
                        type: 'password',
                        onChange: this.handleInputChange
                    }),
                    requiredAttributes.map(function (attribute) {
                        return _react2.default.createElement(_AmplifyUIComponentsReact.Input, {
                            placeholder: _core.I18n.get(convertToPlaceholder(attribute)),
                            theme: theme,
                            key: attribute,
                            name: attribute,
                            type: 'text',
                            onChange: _this4.handleInputChange
                        });
                    })
                ),
                _react2.default.createElement(
                    _AmplifyUIComponentsReact.SectionFooter,
                    { theme: theme },
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterPrimaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Button,
                            { theme: theme, onClick: this.change },
                            _core.I18n.get('Change')
                        )
                    ),
                    _react2.default.createElement(
                        _AmplifyUIComponentsReact.SectionFooterSecondaryContent,
                        { theme: theme },
                        _react2.default.createElement(
                            _AmplifyUIComponentsReact.Link,
                            { theme: theme, onClick: function onClick() {
                                    return _this4.changeState('signIn');
                                } },
                            _core.I18n.get('Back to Sign In')
                        )
                    )
                )
            );
        }
    }]);
    return RequireNewPassword;
}(_AuthPiece3.default);

exports.default = RequireNewPassword;


function convertToPlaceholder(str) {
    return str.split('_').map(function (part) {
        return part.charAt(0).toUpperCase() + part.substr(1).toLowerCase();
    }).join(' ');
}

function objectWithProperties(obj, keys) {
    var target = {};
    for (var key in obj) {
        if (keys.indexOf(key) === -1) {
            continue;
        }
        if (!Object.prototype.hasOwnProperty.call(obj, key)) {
            continue;
        }
        target[key] = obj[key];
    }
    return target;
}
